### ISSUE TITLE
Schema Validations between Teradata and BigQuery ignore nullability constraints

### ISSUE THREAD

**Issue by dominikhron** [GoogleCloudPlatform/professional-services-data-validator#926]
I've created tables in Teradata and BigQuery with various constraints on nullability to check how the Data Validation Tool handles them. The result is that with these two systems, nullability is completely ignored and all columns return a success.

I created a table in **Teradata** like so:

```sql
CREATE TABLE CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY
(
	STRING_MANDATORY VARCHAR(15) NOT NULL,
	STRING_NULLABLE VARCHAR(15),
	STRING_TD_NULL_BQ_MAN VARCHAR(15),
	STRING_TD_MAN_BQ_NULL VARCHAR(15) NOT NULL,
	TIMESTAMP_MANDATORY TIMESTAMP(6) NOT NULL,
	TIMESTAMP_NULLABLE TIMESTAMP(6),
	TIMESTAMP_TD_NULL_BQ_MAN TIMESTAMP(6),
	TIMESTAMP_TD_MAN_BQ_NULL TIMESTAMP(6) NOT NULL,
	DECIMAL_MANDATORY DECIMAL(10,0) NOT NULL,
	DECIMAL_NULLABLE DECIMAL(10,0),
	DECIMAL_TD_NULL_BQ_MAN DECIMAL(10,0),
	DECIMAL_TD_MAN_BQ_NULL DECIMAL(10,0) NOT NULL
)
```

And the following in **BigQuery**:

```sql
create table vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability
(
  string_mandatory string not null,
  string_nullable string,
  string_td_null_bq_man string not null,
  string_td_man_bq_null string,
  timestamp_mandatory timestamp not null,
  timestamp_nullable timestamp,
  timestamp_td_null_bq_man timestamp not null,
  timestamp_td_man_bq_null timestamp,
  decimal_mandatory decimal not null,
  decimal_nullable decimal,
  decimal_td_null_bq_man numeric not null,
  decimal_td_man_bq_null numeric,
)
```

Then I run a **schema validation** like so:

```
data-validation -v validate schema -sc MY_TD_CONN -tc MY_BQ_CONN -tbls CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY=vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability -al "float64:decimal(38,9)"
```

And the **result** of this validation is:

```
╒══════════════════════════════════════╤═══════════════════╤═══════════════════╤══════════╤═════════════════════════════════════╤═════════════════════════════════════════════════╤══════════════════════════╤══════════════════════════╤════════════════════╤════════════════════╤═════════════════════╕
│ run_id                               │ validation_name   │ validation_type   │ labels   │ source_table_name                   │ target_table_name                               │ source_column_name       │ target_column_name       │ source_agg_value   │ target_agg_value   │ validation_status   │
╞══════════════════════════════════════╪═══════════════════╪═══════════════════╪══════════╪═════════════════════════════════════╪═════════════════════════════════════════════════╪══════════════════════════╪══════════════════════════╪════════════════════╪════════════════════╪═════════════════════╡
│ 4153130f-209f-487c-82dc-a8900a717f1b │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ string_mandatory         │ string_mandatory         │ string             │ string             │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 4153130f-209f-487c-82dc-a8900a717f1b │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ string_nullable          │ string_nullable          │ string             │ string             │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 4153130f-209f-487c-82dc-a8900a717f1b │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ string_td_null_bq_man    │ string_td_null_bq_man    │ string             │ string             │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 4153130f-209f-487c-82dc-a8900a717f1b │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ string_td_man_bq_null    │ string_td_man_bq_null    │ string             │ string             │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 4153130f-209f-487c-82dc-a8900a717f1b │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ timestamp_mandatory      │ timestamp_mandatory      │ timestamp          │ timestamp          │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 4153130f-209f-487c-82dc-a8900a717f1b │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ timestamp_nullable       │ timestamp_nullable       │ timestamp          │ timestamp          │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 4153130f-209f-487c-82dc-a8900a717f1b │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ timestamp_td_null_bq_man │ timestamp_td_null_bq_man │ timestamp          │ timestamp          │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 4153130f-209f-487c-82dc-a8900a717f1b │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ timestamp_td_man_bq_null │ timestamp_td_man_bq_null │ timestamp          │ timestamp          │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 4153130f-209f-487c-82dc-a8900a717f1b │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ decimal_mandatory        │ decimal_mandatory        │ float64            │ decimal(38, 9)     │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 4153130f-209f-487c-82dc-a8900a717f1b │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ decimal_nullable         │ decimal_nullable         │ float64            │ decimal(38, 9)     │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 4153130f-209f-487c-82dc-a8900a717f1b │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ decimal_td_null_bq_man   │ decimal_td_null_bq_man   │ float64            │ decimal(38, 9)     │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 4153130f-209f-487c-82dc-a8900a717f1b │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ decimal_td_man_bq_null   │ decimal_td_man_bq_null   │ float64            │ decimal(38, 9)     │ success             │
╘══════════════════════════════════════╧═══════════════════╧═══════════════════╧══════════╧═════════════════════════════════════╧═════════════════════════════════════════════════╧══════════════════════════╧══════════════════════════╧════════════════════╧════════════════════╧═════════════════════╛
```

Do let me know if you need any more information from me. 🙂

**Comment by helensilva14** [comment]
Hi @dominikhron, thanks for pointing this scenario to us, we are assessing it internally and we will get a response back to you soon!

**Comment by nj1973** [comment]
I've added some tests (inspired by your test table @dominikhron) on a feature branch to assess the current situation.

Engines that pick up `nullable=False`: MySQL, Oracle, PostgreSQL, Snowflake, SQL Server
Engines not picking up `nullable=False`: BigQuery, Teradata, Hive, Spanner

We need to give this some thought but at least we now (mostly) understand the engines in scope.

**Comment by nj1973** [comment]
Hive nullable detection is not trivial and to avoid delaying progression of this issue I've spun it off into issue https://github.com/GoogleCloudPlatform/professional-services-data-validator/issues/934

**Comment by nj1973** [comment]
Hi @dominikhron, we've merged some changes into the `develop` branch of this project to hopeful fir the problem you saw.
If you would like re-run your test then please pull the latest changes and give it a try. Please note these new changes are not in the PyPI version, only on github for the time being.
If you have any problems please re-open this issue.

**Comment by dominikhron** [comment]
Thank you @nj1973, I can see the effect of these changes on the develop branch. Here, I must run the following command to get the expected **schema validation** result:

```
data-validation -v validate schema -sc MY_TD_CONN -tc MY_BQ_CONN -tbls CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY=vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability -al "decimal(10,0):decimal(38,9), timestamp:timestamp('UTC')"
```

The **result** of this validation is:

```
╒══════════════════════════════════════╤═══════════════════╤═══════════════════╤══════════╤═════════════════════════════════════╤═════════════════════════════════════════════════╤══════════════════════════╤══════════════════════════╤════════════════════╤════════════════════╤═════════════════════╕
│ run_id                               │ validation_name   │ validation_type   │ labels   │ source_table_name                   │ target_table_name                               │ source_column_name       │ target_column_name       │ source_agg_value   │ target_agg_value   │ validation_status   │
╞══════════════════════════════════════╪═══════════════════╪═══════════════════╪══════════╪═════════════════════════════════════╪═════════════════════════════════════════════════╪══════════════════════════╪══════════════════════════╪════════════════════╪════════════════════╪═════════════════════╡
│ 72dd34b2-9c4e-4e80-85aa-52d4d4bfc3c6 │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ string_mandatory         │ string_mandatory         │ !string            │ !string            │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 72dd34b2-9c4e-4e80-85aa-52d4d4bfc3c6 │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ string_nullable          │ string_nullable          │ string             │ string             │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 72dd34b2-9c4e-4e80-85aa-52d4d4bfc3c6 │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ string_td_null_bq_man    │ string_td_null_bq_man    │ string             │ !string            │ fail                │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 72dd34b2-9c4e-4e80-85aa-52d4d4bfc3c6 │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ string_td_man_bq_null    │ string_td_man_bq_null    │ !string            │ string             │ fail                │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 72dd34b2-9c4e-4e80-85aa-52d4d4bfc3c6 │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ timestamp_mandatory      │ timestamp_mandatory      │ !timestamp         │ !timestamp('UTC')  │ fail                │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 72dd34b2-9c4e-4e80-85aa-52d4d4bfc3c6 │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ timestamp_nullable       │ timestamp_nullable       │ timestamp          │ timestamp('UTC')   │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 72dd34b2-9c4e-4e80-85aa-52d4d4bfc3c6 │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ timestamp_td_null_bq_man │ timestamp_td_null_bq_man │ timestamp          │ !timestamp('UTC')  │ fail                │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 72dd34b2-9c4e-4e80-85aa-52d4d4bfc3c6 │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ timestamp_td_man_bq_null │ timestamp_td_man_bq_null │ !timestamp         │ timestamp('UTC')   │ fail                │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 72dd34b2-9c4e-4e80-85aa-52d4d4bfc3c6 │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ decimal_mandatory        │ decimal_mandatory        │ !decimal(10, 0)    │ !decimal(38, 9)    │ fail                │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 72dd34b2-9c4e-4e80-85aa-52d4d4bfc3c6 │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ decimal_nullable         │ decimal_nullable         │ decimal(10,0)      │ decimal(38, 9)     │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 72dd34b2-9c4e-4e80-85aa-52d4d4bfc3c6 │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ decimal_td_null_bq_man   │ decimal_td_null_bq_man   │ decimal(10, 0)     │ !decimal(38, 9)    │ fail                │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼─────────────────────────────────────┼─────────────────────────────────────────────────┼──────────────────────────┼──────────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 72dd34b2-9c4e-4e80-85aa-52d4d4bfc3c6 │ Schema            │ Schema            │ []       │ CZ_TST_NUCLEUS.DVT_TEST_NULLABILITY │ vfcz_dh_edw_adhoc_dev_01_s.dvt_test_nullability │ decimal_td_man_bq_null   │ decimal_td_man_bq_null   │ !decimal(10, 0)    │ decimal(38, 9)     │ fail                │
╘══════════════════════════════════════╧═══════════════════╧═══════════════════╧══════════╧═════════════════════════════════════╧═════════════════════════════════════════════════╧══════════════════════════╧══════════════════════════╧════════════════════╧════════════════════╧═════════════════════╛
```

I also see this fix has been deployed with PyPI version 4.1.0.

**Comment by helensilva14** [comment]
Hi! Exactly, we released a patch version on Friday so this fix is already available there, you can check the other fixes as well here: https://github.com/GoogleCloudPlatform/professional-services-data-validator/releases/tag/v4.1.0
